Source/CarthageKit/Project.swift

```diff
@@ -587,21 +610,89 @@ public final class Project { // swiftlint:disable:this type_body_length
 		pinnedVersion: PinnedVersion,
 		toolchain: String?
 	) -> SignalProducer<URL, CarthageError> {
+
+		// Helper type
+		typealias SourceURLAndDestinationURL = (frameworkSourceURL: URL, frameworkDestinationURL: URL)
+
+		// Helper type
+		struct GenericError<A>: Error {
+			let value: A
+		}
+
+		// Returns the unique pairs in the input array
+		// or the duplicates keys by .frameworkDestinationURL
+		func uniqueSourceDestinationPairs(
+			_ sourceURLAndDestinationURLpairs: [SourceURLAndDestinationURL]
+			) -> Result<[SourceURLAndDestinationURL], GenericError<[URL: [URL]]>> {
```

Comment:
I would either:

1. Return a `CarthageError`

2. Add a `DuplicatesInArchiveError` type and make `CarthageError`'s case `case duplicatesInArchive(DuplicatesInArchiveError)`

(I'd prefer to move towards the second option, personally)